File: C++/2196. Create Binary Tree From Descriptions.cpp
class Solution {
public:
    void formBinaryTree(TreeNode* root, unordered_map<int, vector<pair<int,pair<int,int>>>> &adj){
        if(adj.find(root->val)==adj.end() || adj[root->val].size()==0) {
            root->left=NULL;
            root->right=NULL;
            return;
        }
        for(auto it:adj[root->val]){
            if(it.second.second==false){
                if(it.second.first==1){
                    root->left=new TreeNode(it.first);
                    it.second.second=true;
                    formBinaryTree(root->left, adj);
                }
                if(it.second.first==0){
                    root->right=new TreeNode(it.first);
                    it.second.second=true;
                    formBinaryTree(root->right, adj);
                }
            }
        }
    }
    TreeNode* createBinaryTree(vector<vector<int>>& descriptions) {
        unordered_map<int, vector<pair<int,pair<int,int>>>> adj;
        unordered_map<int,int> indegrees;
        TreeNode* root = new TreeNode();
        for(int i=0;i<descriptions.size();i++){
            adj[descriptions[i][0]].push_back({descriptions[i][1], {descriptions[i][2], false}});
            indegrees[descriptions[i][1]]++;
        }
        for(auto it: adj){
            if(indegrees.find(it.first)==indegrees.end()){
                root->val=it.first;
            }
        }
        formBinaryTree(root, adj);
        return root;
    }
};

File: JAVA/2196. Create Binary Tree From Descriptions.java
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;    
    TreeNode(int val) {
        this.val = val;
    }
}
class Solution {
    public TreeNode createBinaryTree(int[][] descriptions) {
        Map<Integer, TreeNode> nodeMap = new HashMap<>();
        Set<Integer> children = new HashSet<>();
        for (int[] desc : descriptions) {
            int parentVal = desc[0];
            int childVal = desc[1];
            boolean isLeft = desc[2] == 1;           
            TreeNode parentNode = nodeMap.getOrDefault(parentVal, new TreeNode(parentVal));
            TreeNode childNode = nodeMap.getOrDefault(childVal, new TreeNode(childVal)); 
            if (isLeft) {
                parentNode.left = childNode;
            } else {
                parentNode.right = childNode;
            }            
            nodeMap.put(parentVal, parentNode);
            nodeMap.put(childVal, childNode);            
            children.add(childVal);
        }
        TreeNode root = null;
        for (int val : nodeMap.keySet()) {
            if (!children.contains(val)) {
                root = nodeMap.get(val);
                break;
            }
        }        
        return root;
    }
}
